File: app/workers/export_worker.py
import os
import numpy as np
from PIL import Image
from PySide6.QtCore import QThread, Signal
from app.core.image_document import ImageDocument
from app.processing.compositing import composite_document
from app.utils.logger import logger

class ExportWorker(QThread):
    """
    Asynchronous export worker for rendering and writing high-resolution images to disk.
    Supports PNG (transparent/solid), JPG, WEBP with custom quality and resolution.
    """
    finished = Signal(str)  # Output file path
    error = Signal(str)

    def __init__(self, document: ImageDocument, output_path: str, format_str: str = "PNG", quality: int = 95, width: int = None, height: int = None):
        super().__init__()
        self.doc = document
        self.output_path = output_path
        self.format_str = format_str.upper()
        self.quality = quality
        self.target_width = width
        self.target_height = height

    def run(self):
        try:
            # 1. Render composite image
            comp_array = composite_document(self.doc, preview_mode=False)

            # 2. Resize if target dimensions specified
            if self.target_width and self.target_height:
                pil_img = Image.fromarray(comp_array)
                pil_img = pil_img.resize((self.target_width, self.target_height), Image.Resampling.LANCZOS)
            else:
                pil_img = Image.fromarray(comp_array)

            # 3. Handle JPG conversion (no alpha channel)
            if self.format_str == "JPG" or self.format_str == "JPEG":
                if pil_img.mode == "RGBA":
                    # Replace transparent background with chosen bg color or white
                    bg = Image.new("RGB", pil_img.size, self.doc.bg_color)
                    bg.paste(pil_img, mask=pil_img.split()[3])
                    pil_img = bg
                else:
                    pil_img = pil_img.convert("RGB")
                pil_img.save(self.output_path, "JPEG", quality=self.quality)

            elif self.format_str == "WEBP":
                pil_img.save(self.output_path, "WEBP", quality=self.quality)

            else:
                # Default PNG
                pil_img.save(self.output_path, "PNG", compress_level=6)

            logger.info(f"Successfully exported image to {self.output_path}")
            self.finished.emit(self.output_path)
        except Exception as e:
            logger.error(f"Failed to export image: {e}", exc_info=True)
            self.error.emit(str(e))

File: Cpp/src/workers/ExportWorker.cpp
#include "workers/ExportWorker.h"
#include "processing/Compositor.h"
#include "utils/ImageUtils.h"
#include "utils/Logger.h"

namespace ImageCut {
namespace Workers {

ExportWorker::ExportWorker(
    std::shared_ptr<Core::ImageDocument> document,
    const QString& outputPath,
    const QString& formatStr,
    int quality,
    int width,
    int height
) : m_doc(document),
    m_outputPath(outputPath),
    m_formatStr(formatStr.toUpper()),
    m_quality(quality),
    m_targetWidth(width),
    m_targetHeight(height)
{}

void ExportWorker::run() {
    try {
        if (!m_doc) {
            emit error("Null document reference.");
            return;
        }

        cv::Mat compArray = Processing::Compositor::compositeDocument(*m_doc, false);

        if (m_targetWidth > 0 && m_targetHeight > 0) {
            cv::resize(compArray, compArray, cv::Size(m_targetWidth, m_targetHeight), 0, 0, cv::INTER_LANCZOS4);
        }

        if (m_formatStr == "JPG" || m_formatStr == "JPEG") {
            cv::Mat bg(compArray.rows, compArray.cols, CV_8UC3, cv::Scalar(m_doc->bgColor.red(), m_doc->bgColor.green(), m_doc->bgColor.blue()));
            std::vector<cv::Mat> compChs;
            cv::split(compArray, compChs);

            cv::Mat alphaF;
            compChs[3].convertTo(alphaF, CV_32F, 1.0 / 255.0);

            std::vector<cv::Mat> rgbChs = { compChs[0], compChs[1], compChs[2] };
            cv::Mat fgRgb, fgF, bgF, outF, outRgb;
            cv::merge(rgbChs, fgRgb);

            fgRgb.convertTo(fgF, CV_32F);
            bg.convertTo(bgF, CV_32F);

            std::vector<cv::Mat> alpha3 = { alphaF, alphaF, alphaF };
            cv::Mat a3;
            cv::merge(alpha3, a3);

            outF = fgF.mul(a3) + bgF.mul(1.0 - a3);
            outF.convertTo(outRgb, CV_8UC3);

            Utils::ImageUtils::saveImage(m_outputPath, outRgb, m_quality);
        } else {
            Utils::ImageUtils::saveImage(m_outputPath, compArray, m_quality);
        }

        LOG_INFO("Successfully exported image to: " + m_outputPath.toStdString());
        emit finished(m_outputPath);
    } catch (const std::exception& e) {
        LOG_ERROR(std::string("Failed to export image: ") + e.what());
        emit error(QString::fromStdString(e.what()));
    }
}

} // namespace Workers
} // namespace ImageCut
